feat: Add min_backend_version decorator

AlejandroFernandezLuces opened the pull request:
## Description
Add a minimum version decorator to allow checking if methods are compatible with the current backend.

## Issue linked
#1004

## Checklist
- [x] I have tested my changes locally.
- [x] I have added necessary documentation or updated existing documentation.
- [x] I have followed the coding style guidelines of this project.
- [x] I have added appropriate unit tests.
- [x] I have reviewed my changes before submitting this pull request.
- [x] I have linked the issue or issues that are solved to the PR if any.
- [x] I have assigned this PR to myself.
- [x] I have made sure that the title of my PR follows [Conventional commits style](https://www.conventionalcommits.org/en/v1.0.0/#summary) (e.g. ``feat: extrude circle to cylinder``)

PipKat commented on src/ansys/geometry/core/misc/checks.py:
```diff
@@ -302,3 +303,44 @@ def check_type_all_elements_in_iterable(
+
+def min_backend_version(method_version: str):
+    """
+    Compare the min version required by a method to the current backend version.
```
```suggestion
    Compare a method's minimum required version to the current backend version.
```

PipKat commented on src/ansys/geometry/core/misc/checks.py:
```diff
@@ -302,3 +303,44 @@ def check_type_all_elements_in_iterable(
+    Parameters
+    ----------
+    method_version : str
+        The minimum version required by the method.
```
```suggestion
        Minimum version required by the method.
```

PipKat commented on src/ansys/geometry/core/misc/checks.py:
```diff
@@ -302,3 +303,44 @@ def check_type_all_elements_in_iterable(
+        def wrapper(self, *args, **kwargs):
+            if self._grpc_client is None:
+                raise GeometryRuntimeError(
+                    "The client is not available. Please initialize the client first."
```
```suggestion
                    "The client is not available. You must initialize the client first."
```

RobPasMue commented on .github/pull_request_template.md:
```diff
@@ -12,4 +12,5 @@
 - [ ] I have reviewed my changes before submitting this pull request.
 - [ ] I have linked the issue or issues that are solved to the PR if any.
 - [ ] I have assigned this PR to myself.
+- [ ] I have added the minimum version decorator to any new backend method implemented.
```
Great idea - thanks @AlejandroFernandezLuces

RobPasMue commented on pyproject.toml:
```diff
@@ -38,6 +38,7 @@ dependencies = [
     "scipy>=1.7.3,<2",
     "six>=1.16.0,<2",
     "vtk>=9,<10",
+    "semver>=3.0.2",
```
```suggestion
    "semver>=3,<4",
```
We should block major versions on general deps

RobPasMue replied:
We should also add a pinned version to the "tests" extras

RobPasMue commented on src/ansys/geometry/core/connection/client.py:
```diff
@@ -170,8 +171,9 @@ def __init__(
         self._admin_stub = AdminStub(self._channel)
 
         # if no backend type has been specified, ask the backend which type it is
+        grpc_backend_response = self._admin_stub.GetBackend(Empty())
         if backend_type == None:
```
```suggestion
        # retrieve the backend information
        grpc_backend_response = self._admin_stub.GetBackend(Empty())
        
        # if no backend type has been specified, ask the backend which type it is
        if backend_type == None:
```

RobPasMue commented on src/ansys/geometry/core/connection/client.py:
```diff
@@ -181,6 +183,13 @@ def __init__(
             elif grpc_backend_type == GRPCBackendType.LINUX_DMS:
                 backend_type = BackendType.LINUX_SERVICE
 
+        # backend version
```
```suggestion
        # retrieve the backend version
```

RobPasMue commented on src/ansys/geometry/core/connection/client.py:
```diff
@@ -181,6 +183,13 @@ def __init__(
+            self._backend_version = f"{ver.major_release}.{ver.minor_release}.{ver.service_pack}"
+        else:
+            logger.warning("The backend version is only available after 24.1 version.")
+            self._backend_version = "24.1.0"
         # Store the backend type
```
```suggestion
            self._backend_version = "24.1.0"

        # Store the backend type
```

RobPasMue commented on src/ansys/geometry/core/connection/client.py:
```diff
@@ -202,6 +211,18 @@ def backend_type(self) -> BackendType:
         return self._backend_type
 
+    @property
+    def backend_version(self) -> VersionIdentifier:
```
```suggestion
    def backend_version(self) -> str:
```

RobPasMue commented on src/ansys/geometry/core/connection/client.py:
```diff
@@ -202,6 +211,18 @@ def backend_type(self) -> BackendType:
+
+        Returns
+        -------
+        VersionIdentifier
```
```suggestion
        str
```

RobPasMue commented on src/ansys/geometry/core/connection/client.py:
```diff
@@ -202,6 +211,18 @@ def backend_type(self) -> BackendType:
+        Returns
+        -------
+        VersionIdentifier
+            Backend version.
```
```suggestion
            Backend version in semantic versioning format (that is, Ansys 24R1 SP2 would be ``24.1.2``).
```

RobPasMue commented on src/ansys/geometry/core/misc/checks.py:
```diff
@@ -302,3 +303,44 @@ def check_type_all_elements_in_iterable(
+    from ansys.geometry.core.errors import GeometryRuntimeError
+
+    def backend_version_decorator(method):
+        def wrapper(self, *args, **kwargs):
```
Let's verify that "self" has an attribute called "_grpc_client". It should always be the case but it will also act as a "defensive" measure for developers.

RobPasMue commented on src/ansys/geometry/core/misc/checks.py:
```diff
@@ -302,3 +303,44 @@ def check_type_all_elements_in_iterable(
+
+    def backend_version_decorator(method):
+        def wrapper(self, *args, **kwargs):
+            if self._grpc_client is None:
+                raise GeometryRuntimeError(
```
I'm unsure whether there is any scenario where the client is not defined (as long as the attribute exists). We always pass it in construction to all objects. But in any case, good "defensive" measure

RobPasMue commented on src/ansys/geometry/core/connection/client.py:
```diff
@@ -181,6 +183,13 @@ def __init__(
+            logger.warning("The backend version is only available after 24.1 version.")
+            self._backend_version = "24.1.0"
         # Store the backend type
         self._backend_type = backend_type
```
I would reorder this section so that the "backend_type" logic is all together, and the backend version logic comes right afterwards

RobPasMue reviewed (changes requested):
Overall looking good - some minor changes. Request review again once addressed. Great work @AlejandroFernandezLuces

RobPasMue reviewed (commented):
I am missing some tests by the way. Can you test the decorator on its own? Use monkeypatch to avoid having to adapt the test in the future. Create a random object that has a "_grpc_client" attribute and it also has another attribute called backend version. Testing is required for the decorator only

AlejandroFernandezLuces commented:
Added the test, and fixed some issues that arose from it. Also, it seems that is only working on the unstable version, I guess that is due to the feature not being available in the stable version?

RobPasMue commented:
> Added the test, and fixed some issues that arose from it. Also, it seems that is only working on the unstable version, I guess that is due to the feature not being available in the stable version?

Yep - precisely. This will require a new stable release. Let me trigger it

RobPasMue commented:
Rerunning all jobs - images have been published

AlejandroFernandezLuces commented:
Thanks for the reviews @PipKat @RobPasMue 😄 

Enabling automerge.

RobPasMue commented on pyproject.toml:
```diff
@@ -38,6 +38,7 @@ dependencies = [
     "scipy>=1.7.3,<2",
     "six>=1.16.0,<2",
     "vtk>=9,<10",
+    "semver>=3,<4",
```
Add a pinned version in the tests dependencies

RobPasMue commented on src/ansys/geometry/core/misc/checks.py:
```diff
@@ -302,3 +303,48 @@ def check_type_all_elements_in_iterable(
         check_type(elem, expected_type)
+
+
+def min_backend_version(method_version: str):
```
I'm thinking... why not use 3 integers to define the backend version? major, minor an sp version, And we build the string on our own inside of it.

I'm just thinking of a wrongly-formatted decorator - string input is very flexible. However integers can only be integers. Anyway, what do you think?

AlejandroFernandezLuces replied:
Agree, I prefer your suggestion.

RobPasMue reviewed (changes requested):
Tests look really good. One suggestion and two change requests. After that we can merge

RobPasMue reviewed (commented):
This solves the test issues

RobPasMue commented on tests/test_misc_checks.py:
```diff
@@ -199,3 +201,54 @@ def test_check_ndarray_is_not_none():
+    # If client is not initialized, an error should be raised
+    mock_object._grpc_client = None
+
+    @min_backend_version("24.1.0")
```
```suggestion
    @min_backend_version(24, 1, 0)
```
